Modules/Common/CardanoLikeNativeTokensModule.php

```diff
@@ -0,0 +1,321 @@
+    {
+        $this->version = 1;
+    }
+
+    final public function post_post_initialize()
+    {
+
+        $this->metadata_registry = envm(
+            $this->module,
+            'METADATA_REGISTRY',
+            new DeveloperError('Metadata registry is not set in the config')
+        );
+
+        $this->db = pg_connect($this->select_node());
+
+        $timeout_ms = envm($this->module, 'REQUESTER_TIMEOUT') * 1000;
+        pg_query($this->db, "SET statement_timeout = {$timeout_ms}");
+    }
+
+    final public function inquire_latest_block()
+    {
+        return (int)pg_fetch_assoc(pg_query($this->db, 'SELECT max(block_no) FROM block'))['max'];
+    }
+
+    final public function ensure_block($block_id, $break_on_first = false)
+    {
+        $block = pg_fetch_assoc(pg_query_params($this->db, 'SELECT id, hash, time, tx_count FROM block WHERE block_no = $1', [$block_id]));
+        $this->block_hash = substr($block['hash'], 2);
+        $this->block_time = $block['time'];
+        $this->block_db_id = (int)$block['id'];
+        $this->transaction_count = (int)$block['tx_count'];
+
+        // We don't really check if all our nodes have the same block here as we always use one db connection to
+        // retrieve data within a single block anyways, so this one won't be throwing ConsensusException
+    }
+
+    final public function pre_process_block($block_id)
+    {
+        $events = [];
+        $block = [];
+
+        $transactions = pg_fetch_all(pg_query_params($this->db,
+            'SELECT id, hash, out_sum, fee, deposit FROM tx WHERE block_id = $1 ORDER BY block_index', [$this->block_db_id]));
+        // or this can be done via a subquery like this: WHERE block_id = (SELECT id FROM block WHERE block_no = $1),
+        // and $block_id is the param
+
+        if (count($transactions) !== $this->transaction_count)
+            throw new ConsensusException('Transaction count mismatch (orphaned block?)');
+
+        if (!$transactions)
+        {
+            $this->set_return_events([]);
+            $this->set_return_currencies([]);
+            return;
+        }
+
+        $in_query = [];
+
+        foreach ($transactions as $transaction)
+        {
+            $block[($transaction['id'])] = $transaction;
+            $block[($transaction['id'])]['inputs'] = [];
+            $block[($transaction['id'])]['outputs'] = [];
+            $in_query[] = $transaction['id'];
+        }
+
+        $in_query = implode(', ', $in_query);
+
+        $outs = pg_fetch_all(pg_query($this->db,
+            "SELECT tx_out.tx_id, tx_out.address, multi_asset.fingerprint, ma_tx_out.ident AS token, ma_tx_out.quantity
+                    FROM tx_out
+                    JOIN ma_tx_out ON (ma_tx_out.tx_out_id = tx_out.id)
+                    JOIN multi_asset ON ma_tx_out.ident = multi_asset.id
+                    WHERE tx_out.tx_id in ({$in_query})
+                    ORDER BY tx_out.tx_id, tx_out.index"));
```

Comment:
ordering by ma_tx_out.id should be done as well, because a single output can result in a few input of `ma`:

| ma\_tx\_out.id | tx\_id | id | address | index | fingerprint | token | quantity |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| 718801721 | 74514365 | 209194065 | addr1z8nvjzjeydcn4atcd93aac8allvrpjn7pjr2qsweukpnayd8awyllkdhpwm7r3tnvvm0mzx592fzgazhuevl679cf7lq35qw53 | 0 | asset17tnpuy77fazv5pl2q4u7yxexp34ap7y0nd3z0c | 4135817 | 1 |
| 718801722 | 74514365 | 209194065 | addr1z8nvjzjeydcn4atcd93aac8allvrpjn7pjr2qsweukpnayd8awyllkdhpwm7r3tnvvm0mzx592fzgazhuevl679cf7lq35qw53 | 0 | asset1zmw280a7s3amu5e50d87dem0uduy9mf9rvj9xx | 7656792 | 9223370374484050969 |
| 718801723 | 74514365 | 209194065 | addr1z8nvjzjeydcn4atcd93aac8allvrpjn7pjr2qsweukpnayd8awyllkdhpwm7r3tnvvm0mzx592fzgazhuevl679cf7lq35qw53 | 0 | asset17v9z2sf7v05z6mne4qk0kzlmue5aqxtfxq8jyk | 7650043 | 1769334232149 |
| 718801724 | 74514365 | 209194067 | addr1q87hqhxpdat6du2m4tzgkrftn89zdnmu2ql2fpmuwqe85zs4mrjqlk30ekjfdanrl04k6fc4xnhmnpz9lnmhge8wa77s7p0ukd | 2 | asset1edg0amsv72pks7uz8r8m26e4ljweprk46tneud | 4075427 | 1 |